src/IO/ReadBufferFromS3.cpp

```diff
@@ -351,8 +380,27 @@ bool ReadBufferFromS3::atEndOfRequestedRangeGuess()
     return false;
 }
 
+ReadBufferFromS3::~ReadBufferFromS3()
+{
+    try
+    {
+        if (!read_all_range_successfully && read_result)
+            /// When we abandon a session with an ongoing GetObject request and there is another one trying to delete the same object this delete
+            /// operation will hang until GetObject's session idle timeouts. So we have to call `reset()` on GetObject's session session immediately.
+            resetSession(*read_result);
```

Comment:
is there a big overhead if resetting the session regardless of `read_all_range_successfully`?

Reply:
it means close connection + reconnect on the next request. so it would eliminate whole purpose of connection pool.